README.md

```diff
@@ -145,12 +142,93 @@ for a link-local address is auto-detected.
 
 ## Detailed Description of Capabilities
 
-Here are some more detailes on the additional capabilities of mavlink-router.
-
-### Endpoint Configuration
-
-A description of the different endpoint operating modes and their configuration
-parameters can be found in the [examples/config.sample](examples/config.sample).
+To understand how MAVLink Router forwards messages between endpoints, it
+important to know, that a MAVLink message has to have a sender address, but
+only some messages have a target address (e.g. parameter requests). The sender
+and target address consist of a system and component ID each, so one UAV can
+have a flight controller as well as other services running on a companion
+computer with individual component IDs, but sharing the same system ID. In the
+target address, a component ID of 0 is used to broadcast to all components on a
+system, a sysID of 0 broadcasts to all systems.
+
+### Endpoints
+MAVLink Router supports three basic types of endpoints: UART, UDP link and TCP
+client. Additionally, it'll act as a TCP server for dynamic clients (if not
+explicitly deactivated).
+
+Endpoint types (see [examples/config.sample](examples/config.sample) for the
+config file format):
+
+  - UART: For telemetry radios or other serial links
+    * Configuration: UART device path/name and baudrate
+    * Behavior: Data is received and sent without waiting for incoming data first
+  - UDP:
+    * Configuration: Mode (client or server), IP address and port; both unicast
+      and multicast addresses are possible
+    * Behavior in client mode: Endpoint is configured with a target IP and port
+      combination. So MAVLink messages can be sent directly after startup, but
+      will only be recevied after the first message was received by the remote
+      side, it doesn't know our IP and port otherwise.  
+      When using a broadcast or multicast IP address, messages will always be
+      sent to that broadcast/multicast IP.
```

Comment:
@lucasdemarchi Is that how it should work with UDP client and a broadcast IP address? Or do we really need to start with sending IP broadcasts, switch to a unicast connection after the first incoming packet and switch back to broadcast after a timeout.

Reply:
current behavior when using a broadcast address is that it will send to that address until it receives a reply. When we receive a reply from, say, 192.168.0.10:54243  we "lock in this address:port" and all new packets are sent there.
We never switch back to broadcast, but it would be good if we did: otherwise if the connection is broken we never recover.

Reply:
The current behavior is clear, but it's quite clearly wrong (issue exists) and therefore I would like to describe the way it should be in this PR.
The question is, if it should be fixed to broadcast forever or whether we really need that timeout.

Reply:
Broadcast forever would be wrong. We need a timeout.

Reply:
All systems should be sending a hearbeat, so we should receive at least 1 message every second or so. A 5s timeout would be reasonable.

Reply:
OK. But I would propose a third mode for that in this case. Because it behaves quite differently than all the other modes. What do you think?

Reply:
No, I think not going back to broadcast was a missing feature. It's only triggered if the address supplied by the user is a broadcast address.

Reply:
So no new mode for the UDP endpoint but rather some special behavior triggered by a broadcast IPv4 address? But how to reliably detect a broadcast address without knowing the subnet size? We can't just assume everyone uses /24 networks.

Reply:
If we needed, we could query it from the kernel. But just setting the broadcast flag regardless if it's a broadcast address proved to work: it will be ignored by the kernel if it's not a broadcast address.

Reply:
But how do we determine then (just with the existing settings), if we need to check for a timeout or not? Or do we just always check for a timeout in client-mode regardless if the address is broadcast or not?
When thinking about the latter option: It should be any different than currently for unicast addresses, because the source address of the latest incoming packet should be the same as the configured one. So resetting to the configured address after a timeout has no effect. Is that what you were thinking of?

Reply:
We need to keep a copy of the address that was configured by the user and switch back to it after a disconnect timeout.
If it was not a broadcast address, nothing will actually change, because recvfrom() won't have changed that.

Supposing the following scenario:
mavlink-router: inet 192.168.2.42/24 brd 192.168.2.255
GCS: inet 192.168.2.1/24 brd 192.168.2.255

Endpoint configuration: 192.168.2.255:14550

So when we start mavlink-router we will start sending the packets to 192.168.2.255 (broadcast). GCS will receive with src_address=192.168.2.42:<something>, so it responds to that IP:port. When that happens, mavlink-router will receive a packet from, recvfrom() in mavlink-router will receive a packet from 192.168.2.1:14550. Currently we are overriding the configured address and losing it - we should rather save it.  All subsequent packets we send through the Endpoint we are then sending to 192.168.2.1:14550.  Proposal here is that when a timeout happens, we handle the "disconnect" by switching back to 192.168.2.255:14550.